### PR TITLE
[Narwhal] Add a `min_header_delay` parameter

### PR DESCRIPTION
**Pull request by mwtian**
Currently we are trying to use the header batches threshold to determine the speed of generating headers, which depends on the rate of batch creations, which in turn depends on the write rate. Instead, adding a new `min_header_delay` parameter, which allows proposing headers before reaching batch threshold, would make the header propose latency more predictable especially under lower TPS.

For now, the `min_header_delay` is set to 1.8s, only slightly below `max_header_delay`, to separate behavior change from this PR. We will run experiments on different values. One possibility as @arun-koshy proposed is to set `min_header_delay` to 1s and `max_header_delay` to 5s.

Also, ensure leader only needs to wait for parent stakes and nothing else.

### CODE REVIEW

**Review by asonnino — APPROVED** [review]
Nice!

File: narwhal/primary/src/proposer.rs [hunk 1]
```diff
@@ -304,20 +308,35 @@ impl Proposer {
...
+
+    fn min_delay(&self) -> Duration {
+        match self.network_model {
+            // In partial synchrony, if this node is going to be the leader of the next
+            // round, we use a lower min delay value to increase its chance of committing
+            // the leader.
+            NetworkModel::PartiallySynchronous
+                if self.committee.leader(self.round + 1) == self.name =>
+            {
+                Duration::ZERO
```
Comment (LefKok) [inline narwhal/primary/src/proposer.rs:335]: cool :)

File: narwhal/primary/src/proposer.rs [hunk 2]
```diff
@@ -592,9 +629,12 @@ impl Proposer {
                     let _ = ack_channel.send(());
                 }
 
-                // Check whether the timer expired.
-                () = &mut timer, if !timer_expired => {
-                    // Nothing to do.
+                // Check whether any timer expired.
+                () = &mut max_delay_timer, if !max_delay_timed_out => {
+                    // Continue to next ieration of the loop.
```
Comment (arun-koshy) [inline narwhal/primary/src/proposer.rs:634]: s/ieration/iteration (same in the below comment)

**Review by akichidis — APPROVED** [review]
Great addition! 👍

**Comment by mwtian** [comment]
> Awesome! Just curious what are the effects to dag performance if we hit the min timeout eventually and add empty headers?

If this happens, the network must have very low load. Generating headers at 1/s ~ 2/s should be ok. And the benefit would be lower latencies.

**Comment by arun-koshy** [comment]
> 
> If this happens, the network must have very low load. Generating headers at 1/s ~ 2/s should be ok. And the benefit would be lower latencies.

Spoke with Mingwei offline about this, but in the case we do under tune this parameter it is possible we will generate empty headers. Just something to monitor for the future

**Pull request merged**